Add tutorials guidelines in the main tutorials branch (#1950)
ansys/pydpf-core: doc/source/getting_started/write_doc/guide_lines_tutorials.rst

Reviewer: please rename file as `guidelines_tutorials.rst`

> .. _ref_guide_lines_tutorials:

Reviewer: ```suggestion
.. _ref_guidelines_tutorials:
```

> ============================
> Writing tutorials guidelines
> ============================

Reviewer: ```suggestion
=================
Writing tutorials
=================
```

> You can improve the Py-DPF-Core documentation by adding a:

Reviewer: ```suggestion
You can improve the PyDPF-Core documentation by adding a:
```

> - :ref:`New tutorials section<ref_guide_lines_add_new_tutorial_section>`;
> - :ref:`New tutorial<ref_guide_lines_add_new_tutorial>`.

Reviewer: ```suggestion
- :ref:`New tutorials section<ref_guidelines_add_new_tutorial_section>`;
- :ref:`New tutorial<ref_guidelines_add_new_tutorial>`.
```

> To do so, you must follow the guide lines presented here.

Reviewer: ```suggestion
To do so, you must follow the guidelines presented here.
```

You also need to understand the structure of the ``doc`` directory on the PyDPF-Core library:

.. code-block::

    .
    ├── doc
    │   ├── source
    │   │    ├── api
    │   │    ├── examples
    │   │    ├── getting_started
    │   │    ├── images
    │   │    ├── user_guide
    │   │    ├── conf.py
    │   │    ├── index.rst
    │   ├── styles
    │   ├── make.bat

> You will be handling only the ``doc/source/user_guide`` directory .

Reviewer: ```suggestion
Tutorials are located in the ``doc/source/user_guide`` directory.
```

> .. _ref_guide_lines_add_new_tutorial_section:

Reviewer: ```suggestion
.. _ref_guidelines_add_new_tutorial_section:
```

=============================
Adding a new tutorial section
=============================

:download:`Download the new tutorial section template<tutorial_section_template.rst>`

.. note::

>     Avoid creating new folders unless it is absolutely necessary. If you are in doubt, its precise location can be
>     advised on the pull request. If you must create a new folder, make sure to add a ``index.rst`` file containing
>     a reference, a title and a description of the section. Otherwise the new folder will be ignored by Sphinx.

Reviewer: ```suggestion
    Avoid creating new folders unless absolutely necessary. When in doubt, mention the location of the new section in the pull request for approval. If you must create a new folder, make sure to add an ``index.rst`` file with a reference, a title, and a description of the section. The documentation ignores folders lacking this file.
```

Location and naming
-------------------

> The new tutorial section must be organized in a new folder in ``doc/source/user_guide/tutorials/new_section_name``.

Reviewer: ```suggestion
The new tutorial section resides in a new folder in ``doc/source/user_guide/tutorials/new_section_name``.
```

.. code-block::

    .
    ├── doc
    │   ├── source
    │   │    ├── user_guide
    │   │    │   ├── tutorials
    │   │    │        ├── new_section

Structure
---------

> The new folder must contain at least a ``index.rst`` file. This file has:

Reviewer: ```suggestion
The section folder must contain an ``index.rst`` file with:
```

> - Reference name;
> - Section title;
> - General description of the content of the tutorials in this section;
> - Cards with the tutorial title, description and applicable solvers (the card must have a link to the tutorial file);
> - Toctree with the tutorials in the section.

Reviewer: ```suggestion
- a reference tag for referencing this section in other parts of the documentation
- a title for the tutorial section
- a general description of the theme for the tutorials in this section
- cards with links to the tutorials, titles, descriptions and applicable solvers
- a ``Toctree`` for the tutorials in the section to appear in the navigation pane
```

Author: ```suggestion
- a reference tag for referencing this section in other parts of the documentation,
- a title for the tutorial section,
- a general description of the topics covered in the tutorials in this section,
- cards with links to the tutorials, titles, descriptions and applicable solvers,
- a ``Toctree`` for the tutorials in the section to appear in the navigation pane.
```

.. code-block::

    .. _ref_tutorial_new_section_template:

    =============
    Section title
    =============

    These tutorials demonstrate how to ...

    .. grid:: 1 1 3 3
        :gutter: 2
        :padding: 2
        :margin: 2

        .. grid-item-card:: Tutorial title
           :link: ref
           :link-type: ref
           :text-align: center

           This tutorial ...

           +++
           :bdg-mapdl:`MAPDL` :bdg-lsdyna:`LS-DYNA` :bdg-fluent:`FLUENT` :bdg-cfx:`CFX`

    .. toctree::
        :maxdepth: 2
        :hidden:

>         tutorial_file.rst

Reviewer: @luisaFelixSalles is it possible to directly do an include of the `tutorial_section_template.rst`?

Author: I'll check

Author: Yes it is. I used the literalinclude directive

> You must also add the new section ``index.rst`` file in the main user guide page toctree. You can find this toctree

Author: ```suggestion
You must reference the new section ``index.rst`` file in the main user guide page toctree for it to appear in the sidebar of the user guide main page. You can find this toctree
```

> at the end of ``doc/source/user_guide/index.rst`` file.

Reviewer: ```suggestion
at the end of the ``doc/source/user_guide/index.rst`` file.
```

.. code-block::

    .. toctree::
        :maxdepth: 2
        :hidden:
        :caption: Tutorials

        tutorials/data_structures/index.rst
        tutorials/language_and_usage/index.rst
        tutorials/post_processing_basics/index.rst
        tutorials/import_data/index.rst
        tutorials/mesh/index.rst
        tutorials/transform_data/index.rst
        tutorials/export_data/index.rst
        tutorials/plot/index.rst
        tutorials/animate/index.rst
        tutorials/enriching_dpf_capabilities/index.rst
        tutorials/distributed_files/index.rst
        tutorials/dpf_server/index.rst
        tutorials/licensing/index.rst

>         tutorials/new_section/index.rst

Reviewer: Not sure we want to hardcode this here as it may be subject to change. I know this is only an illustration but I am not sure it is necessary and whether it will actually help.

> .. _ref_guide_lines_add_new_tutorial:

Reviewer: ```suggestion
.. _ref_guidelines_add_new_tutorial:
```

=====================
Adding a new tutorial
=====================

> :download:`Download the tutorial card template<tutorial_card_template.rst>` :download:`Download the tutorial structure template<tutorial_structure_template.rst>`

Reviewer: Isn't that already provided in the section template?

Reviewer: I think it would be simpler to just mention having to add a `.. grid-item-card::` in the section? The card template is not usable as-is since it also includes the grid definition, no?

:download:`Download the tutorial content formating template<tutorial_content_template.rst>`

Location and naming
-------------------

> New tutorials must be added as ``.rst`` files to: ``doc/source/user_guide/tutorials/section_name/tutorial_file.rst``

Reviewer: ```suggestion
New tutorials correspond to new ``.rst`` files in tutorial section folders, for example: ``doc/source/user_guide/tutorials/section_name/tutorial_file.rst``
```

.. code-block::

    .
    ├── doc
    │   ├── source
    │   │    ├── user_guide
    │   │    │   ├── tutorials
    │   │    │        ├── section
    │   │    │             ├── new_tutorial.rst

> You also have to add it to a card and the toctree on the tutorial section ``index.rst`` file. The card must have:

Reviewer: ```suggestion
You must also add a new card in the ``index.rst`` file for the tutorial section as well as modify its toctree. The card must include:
```

> - Tutorial title;
> - Short description;
> - Badges with the applicable solvers;
> - Link to the tutorial file;

Reviewer: ```suggestion
- a tutorial title
- a short description
- badges for the applicable solvers
- a link to the tutorial file
```

Reviewer: @luisaFelixSalles actually I do not the `link to the tutorial file` in the card example

.. topic:: Card example

    .. card:: Tutorial title
       :text-align: center
       :width: 25%

       Short description of the tutorial

       +++

>        :bdg-mapdl:`MAPDL` :bdg-lsdyna:`LS-DYNA` :bdg-fluent:`FLUENT` :bdg-cfx:`CFX`

Reviewer: Same thing as for the section, could we directly do an `include` of the template rst?

Reviewer: ```
.. literalinclude:: tutorial_card_template.rst
   :language: rst
```

Structure
---------

> The tutorial structure can be divided in two main parts:

Reviewer: ```suggestion
The tutorial is divided in two main parts:
```

> - :ref:`Preamble<ref_guide_lines_tutorial_preamble>`;
> - :ref:`Content<ref_guide_lines_tutorial_content>`.

Reviewer: ```suggestion
- :ref:`Preamble<ref_guidelines_tutorial_preamble>`;
- :ref:`Content<ref_guidelines_tutorial_content>`.
```

Author: ```suggestion
- :ref:`Preamble<ref_guidelines_tutorial_header>`;
- :ref:`Content<ref_guidelines_tutorial_content>`.
```

> .. _ref_guide_lines_tutorial_preamble:

Reviewer: ```suggestion
.. _ref_guidelines_tutorial_preamble:
```

Author: ```suggestion
.. _ref_guide_lines_tutorial_header:
```

> Preamble

Reviewer: ```suggestion
Introduction
```
I think `preamble` is not commonly used. A simpler synonym here would be `introduction`?

^^^^^^^^

This first part is essential for clarity, organization and usability of the tutorial. It establishes the tutorials
purpose, making it easy to understand what is going to be explained and reference it within the other parts of
the documentation.

> The preamble must have the following components:

Author: ```suggestion
The header must have the following components:
```

> - File reference name;
> - Tutorial title;
> - Substitution text for the PyDPF-Core library references that will be used across the tutorial;
> - Short description (same phrase used in the tutorial card in the tutorial section ``index.rst`` file);
> - Introduction that explains the context of the tutorial;
> - Download script buttons;

Reviewer: ```suggestion
- a reference tag
- a tutorial title
- any substitution text for references to the PyDPF-Core library used in the tutorial
- a short description (same as for the tutorial card in the tutorial section)
- an introduction
- download buttons for Python script and Jupyter notebook versions of the tutorial
```

.. code-block::

    .. _ref_tutorial_template:

    ==============
    Tutorial title
    ==============

    .. |Examples| replace:: :class:`ansys.dpf.core.examples`

    This sentence resumes the goal of the tutorial

    Introduction to the tutorial

>     :jupyter-download-script:`Download tutorial as Python script<file_name>` :jupyter-download-notebook:`Download tutorial as notebook<file_name>`

Author: ```suggestion
    :jupyter-download-script:`Download tutorial as Python script<file_name>` :jupyter-download-notebook:`Download tutorial as Jupyter notebook<file_name>`
```

The main PyDPF-Core library references are available already defined in the ``doc/source/links_and_refs.rst`` file.
To employ them, you use the ``include`` directive and use the substitution text as usual:

.. code-block::

    .. _ref_tutorial_template:

    ==============
    Tutorial title
    ==============

    .. include:: ../../../links_and_refs.rst

    Here some text. Here we use the |MeshedRegion| substitution text

For more information on those references check the :download:`links and references file<../../links_and_refs.rst>`.

> .. _ref_guide_lines_tutorial_content:

Reviewer: ```suggestion
.. _ref_guidelines_tutorial_content:
```
